docusaurus/docs/recipes/contentful.md

```diff
@@ -0,0 +1,520 @@
+    const middlewares = applyMiddleware(thunkMiddleware.withExtraArgument({ client }));
+
+    return createStore(reducer, initialState, middlewares);
+    }
+}
+
+export default buildStore;
+```
+
+Then, in your actions scripts, you can use the provided client to fetch the content that you need, like so:
+
+```js
+// usecases/actions.js
+// Create a fetch action that fetches data from Contentful
+const fetch = () => async (dispatch, getState, { client }) => {
+    dispatch({ type: 'Fetching'})
+
+    try {
+        // Fetch content that are instances of `useCase` content model, in Spanish
+        const result = await client.getEntries({
+            content_type: 'useCase',
+            locale: 'es-ES',
+        })
+
+        // The incoming object will be fairly complex, and it's recommended to build
+        // a more conveniently structured object here so that other scrips in the pipeline
+        // can avoid handling that complexity and can instead enjoy a simpler, custom built object.
+        // E.G.
+        const entries = result.items
+            .map((item)=> ({
+                title: item.fields.title,
+                description: item.fields.description,
+                weight: item.fields.weight,
+        })).sort((a, b) => b.weight - a.weight);
+
+        dispatch({
+            type: 'Success',
+            payload: { entries },
+        })
+    } catch (error) {
+        dispatch({
+            type: 'Failure',
+            payload: { error },
+        });
+    }
+};
+```
+
+### 3. Conditionally rendering DOM elements in case of preview
+
+Since we're on the topic, you must use a similar strategy in your `App.js` file to conditionally render a DOM element to let the user know they are in a preview state. This must happen once per instance of the App, since we will not perpetuate the `cms-preview` query parameter on route changes.
+
+Class component:
+
+```js
+// App.js
+class App extends Nextapp {
+    state = {
+        isPreviewingContentful = false,
+    };
+
+    // Using the componentWillMount() lifecycle function guarantees that this runs only once per instance of App
+    componentWillMount() {
+        const { router } = this.props;
+
+        this.setState({
+            isPreviewingContentful: Object.hasOwnProperty.call(router.query, 'cms-preview');
+        });
+    }
+
+    render() {
+        return (
+            (...)
+                { this.isPreviewingContentful && <ContentfulPreview> }
+            (...)
+        );
+    }
+}
+```
+
+Using hooks: 
+
+```js
+// App.js
+const App = ({ Component, pageProps, rootSelector, router }) => {
+    const [isPreviewingContentful, setIsPreviewingContentful] = useState(false);
+    
+    // Using a useEffect hook with no dependencies guaranties that it fires only once per instance of App
+    useEffect(() => setIsPreviewingContentful(Object.hasOwnProperty.call(router.query, 'cms-preview')), [])
```

Comment:
```suggestion
    useEffect(() => setIsPreviewingContentful(Object.hasOwnProperty.call(router.query, 'cms-preview')), [router, setIsPreviewingContentful])
```

Reply:
@acostalima, as mentioned in the comment, this `useEffect` must have no dependencies so that it doesn't memoize any object that can change, this in effect guarantees that it only fires once.

This does break one of our current eslint rules though.

Also, huge thanks @dominguesgm for working on my mistakes 🙏

Reply:
Which comment? Sorry, I'm kind of lost. 🙏

Reply:
We can ignore `setIsPreviewingContentful` since its reference never changes anyway.

Reply:
Can `router` change after App is mounted?

Reply:
@acostalima, you are correct, we can set `router` and `setIsrPreviewingContentful`  as dependencies. My original intent was for developers not to memoize `router.query` specifically, which auto-fix in `eslint` does automatically (i.e., if you leave the dependency array empty, and run auto-fix, `eslint` will set `router.query` as a dependency).

I think the documentation should reflect this, but i'm willing to do a PR for this after this first merge if more discussion is warranted.

Reply:
I've never came across a situation where I was using an `useEffect` referencing props or state without specifying its deps. In this case, as a general rule, we should always specify the deps as best practices suggest.